{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1CO8zw4g6g8I2YT7Et0R0mY4cWpeMg-v3","timestamp":1711300707800},{"file_id":"1bUDcgWr6vSQLioLfWtMIA15xpdKYZuUN","timestamp":1706036000814}],"toc_visible":true,"collapsed_sections":["vQlf-NM5D61W","nQ11gmfct4lo"]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# **Mission of clustering notebook**\n","\n","You are a member of a bioinformatics team investigating the [effect of spaceflight on astronaut health](https://en.wikipedia.org/wiki/Effect_of_spaceflight_on_the_human_body).  Your team is trying to find the [biological pathways](https://en.wikipedia.org/wiki/Biological_pathway) that respond to exposure of adverse conditions in space such as [microgravity](https://www.nasa.gov/centers-and-facilities/glenn/what-is-microgravity/) and [radiation](https://www.nasa.gov/directorates/esdmd/hhp/space-radiation/).  Ultimately your team will need phenotype data as labels to train your supervised learning models, but for now, all you have is the gene expression data in the form of [RNA-seq](https://en.wikipedia.org/wiki/RNA-Seq).  \n","\n","Your mission is to use unsupervised machine learning - [clustering](https://en.wikipedia.org/wiki/Cluster_analysis) - to determine if the astronaut samples and their ground control counterparts have clearly distinguishable gene expression profiles.  For if they do, then there is a chance that a supervised learning method can predict responses to spaceflight using different combinations and weights of gene expression from the RNA-seq data.\n","\n","In this notebook, you will use k-means, PCA, and a heatmap to reach your conclusions."],"metadata":{"id":"QG3X-tGIpkMA"}},{"cell_type":"markdown","source":["# Read in the methods\n","\n","**IMPORTANT**: Make sure you put a copy of the `methods.ipynb` in your google drive by following [these instructions](https://docs.google.com/document/d/1V9a3Z5YKT2Pbef4fgPAwB83bHX-p-rPBRRwo7w5Bi9k/edit?usp=sharing).\n","\n","We need to read those methods into this notebook so that we can use them here.  You will get prompted to select the gmail address to use to permit access to your google drive for this notebook.\n","\n","Note that we will import the methods in the notebook as \"m\", so all subsequent references to methods in that notebook will be prefixed with \"m.\"."],"metadata":{"id":"vQlf-NM5D61W"}},{"cell_type":"code","source":["# install and import the python module for importing a notebook\n","!pip install import_ipynb\n","import import_ipynb"],"metadata":{"id":"hCg-4en9D_U2"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# mount your google drive to this notebook\n","from google.colab import drive\n","drive.flush_and_unmount()\n","drive.mount(\"mnt\", force_remount=True)"],"metadata":{"id":"QWwLdzUfHLzL"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# import the \"Copy of methods.ipynb\" from your google drive into this notebook (this will take a while -- mabye 5 minutes?)\n","m = __import__(\"mnt/MyDrive/Colab Notebooks/Copy of methods\")"],"metadata":{"id":"krpbr4EYFvBq"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# read in data and metadata\n","\n","The data that we will be using is [normalized](https://en.wikipedia.org/wiki/Normalization_(statistics)) RNA-seq data which was generated from retinal tissue.  The counts of genes per sample are represented in a table and have been normalized, which means they've been changed to account for differences in how the RNA-seq experiment played out.  For example, some transcripts may have been copied multiple times in the experiment (called [sequencing depth](https://www.biostars.org/p/282708/)) while others may not have been as deeply copied.  Moreover, because some genes are much longer than others, longer genes will have more transcript fragments and appear to be more highly expressed than shorter genes.  These differences will change the results of our analysis because we are looking for differences in gene expression (i.e. counts) between genes and between samples.\n"," Normalization is a statistical process of correcting for these differences, and there are many methods available.  The goal of normalization is to make the gene expression profiles more comparable across samples, allowing for accurate comparisons and statistical analysis."],"metadata":{"id":"nQ11gmfct4lo"}},{"cell_type":"code","source":["# read in the RNA-seq data from OSD-255\n","data=dict()\n","metadata=dict()\n","data['255-normalized'] = m.read_rnaseq_data('255_rna_seq_Normalized_Counts')\n","metadata['255'] = m.read_meta_data('255')"],"metadata":{"id":"zOZ6lCGXt7ht"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# display the dimensions of the RNA-seq data set and the associated metadata\n","print('data shape: ', data['255-normalized'].shape)\n","print('metadata shape: ', metadata['255'].shape)"],"metadata":{"id":"tAHib1PETJyj"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Use the K-means algorithm to cluster the RNA-seq data"],"metadata":{"id":"i6dqhtzx6Cjh"}},{"cell_type":"markdown","source":["## cluster using original unfiltered data"],"metadata":{"id":"De8hR3SamQHV"}},{"cell_type":"code","source":["# make a copy of the original rna-seq dataframe\n","df = data['255-normalized']"],"metadata":{"id":"KJkoKl1bmAAk"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# run the my_kmeans method with k=2\n","m.my_kmeans(df, metadata['255'], k=2)"],"metadata":{"id":"Sg5tcV5imGF7"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# find ideal value of k in k-means for this data\n","m.find_k_elbow(df)"],"metadata":{"id":"oHXP-iy9qgoB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# run the my_kmeans method with k=3\n","m.my_kmeans(df, metadata['255'], k=3)"],"metadata":{"id":"AKZnXy6tqqL_"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## cluster using custom-filtered data"],"metadata":{"id":"Z575BbCEq3GD"}},{"cell_type":"code","source":["# filter out nans and remove genes with CV < 0.1\n","print('before filter: ', df.shape)\n","df = m.filter_data(df, dropnans=True, dropgenes=None, droplowcvs=0.1)\n","print('after filter: ', df.shape)"],"metadata":{"id":"hAEcHvAWq5mb"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# find ideal value of k in k-means for this data\n","m.find_k_elbow(df)"],"metadata":{"id":"yCX_wDEgq_fA"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# run the my_kmeans method with k=4\n","m.my_kmeans(df, metadata['255'], k=4)"],"metadata":{"id":"XrbMJI3qrAOa"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## cluster using DGEA-filtered data"],"metadata":{"id":"vdSLaCqFmVtp"}},{"cell_type":"code","source":["# filter genes to those significantly differentially expressed between ground control and space flight\n","df = m.filter_by_dgea(data['255-normalized'], metadata['255'],  pval=0.05, l2fc=0)"],"metadata":{"id":"7-MIC5peM-4t"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# find ideal value of k in k-means for this data\n","m.find_k_elbow(df)"],"metadata":{"id":"LLuwnCyh0mNv"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# run the my_kmeans method with k=2\n","m.my_kmeans(df, metadata['255'], k=2)"],"metadata":{"id":"xH6v_Y_AiTAZ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["It looks like the gene expression of all the ground control samples clusters uniquely into group 1.   If we insist on 3 clusters, then sample GSM3932708 gets put into its own cluster, suggesting that may be an outlier in the spaceflight group. What we need is a visual way to view how the samples cluster and separate between ground control and spaceflight.  For that we turn to PCA clustering."],"metadata":{"id":"PLfPT1E5gY7N"}},{"cell_type":"markdown","source":["##**QUESTIONS**\n","\n","1. Using the histogram of CVs, about how many genes have a CV of 0.1 or less?  \n","\n","2. Based on these results, what is the ideal value of k for k-means for this data?\n","\n","3. Given these results, would you say that spaceflight and ground control samples have distinct gene expression profiles?"],"metadata":{"id":"UfjUdNMGaoTw"}},{"cell_type":"markdown","source":["**Double click here to enter your answers to the questions above.**\n","\n","1.\n","\n","2.\n","\n","3."],"metadata":{"id":"2iv1T5CwRsGi"}},{"cell_type":"markdown","source":["# Use PCA to cluster and plot the RNA-seq data\n","\n","Let's use the PCA clustering technique so we can visualize the data in 2-dimensions to see how spaceflight and ground control samples cluster and separate."],"metadata":{"id":"ffLp_aCH6PMS"}},{"cell_type":"code","source":["# filter genes to those significantly differentially expressed between ground control and space flight\n","df = m.filter_by_dgea(data['255-normalized'], metadata['255'],  pval=0.05, l2fc=0)"],"metadata":{"id":"j8Pobw-D2QJ4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# transpose the dataframe in preparation for PCA\n","print('shape before transpose: ', df.shape)\n","X = df.drop(columns=['Unnamed: 0']).to_numpy().T\n","print('shape after transpose: ', X.shape)\n"],"metadata":{"id":"a3x8eJTIdS-3"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# create an array y that represents whether the sample is ground control or spaceflight\n","y = m.np.array(list(metadata['255']['Factor Value[Spaceflight]']))\n","target_names=m.np.array(['Ground Control', 'Space Flight'])\n","\n","# run PCA to reduce dimensions from 23,419 to 2!\n","pca = m.PCA(n_components=2)\n","X_r = pca.fit_transform(X)\n","\n","# Percentage of variance explained for each components\n","print(\n","    \"explained variance ratio (first two components): %s\"\n","    % str(pca.explained_variance_ratio_)\n",")\n","\n","# plot the pca plot\n","m.plt.figure()\n","colors = [\"navy\", \"turquoise\"]\n","lw = 2\n","for color, i, target_name in zip(colors, ['Ground Control', 'Space Flight'], target_names):\n","    m.plt.scatter(X_r[y == i, 0], X_r[y == i, 1], color=color, alpha=0.8, lw=lw, label=target_name)\n","m.plt.legend(loc=\"best\", shadow=False, scatterpoints=1)\n","m.plt.title(\"PCA of OSD-255 dataset\")\n","\n","# show the last 3 digits of the sample name\n","samples = list(df.columns)[1:]\n","for i, txt in enumerate(samples):\n","    m.plt.annotate(txt[-3:], (X_r[i][0], X_r[i][1]))\n","m.plt.show()\n"],"metadata":{"id":"qJWkkrMo6RQY"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":[" We're not using the PCA plot for modeling the data and making predictions so there's no real concept of accuracy or performance.  Instead, we're using it to get a visual representation of the distribution of the data.  Now  we see how the ground control and space flight samples cluster together except for a few outliers."],"metadata":{"id":"vtmWb4_N8BWf"}},{"cell_type":"markdown","source":["**QUESTIONS**\n","\n","1. Which samples are the outlier space flight samples?\n","\n","2. Which samples are the outlier ground control samples?\n","\n","3. Do these PCA clusters corroborate what we found using K-means in terms of data separation and outliers?"],"metadata":{"id":"TxHINVq7dwmW"}},{"cell_type":"markdown","source":["**Double click here to enter your answers to the questions above.**\n","\n","1.\n","\n","2.\n","\n","3."],"metadata":{"id":"1XhxYdmNRyK0"}},{"cell_type":"markdown","source":["# Use the DESeq2 implementation of PCA\n","\n","[DESeq2](https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) is a software implementation that performs [differential gene expression analysis](https://en.wikipedia.org/wiki/Gene_expression_profiling) to identify which genes are differentially expressed (highly up-regulated or down-regulated) relative to other genes in the expression profile.  This tool is used to identify which genes may be responsible for a particular phenotype.  \n","\n","In addition to performing differential gene expression analysis, DESeq2 also has an implementation of PCA.  This section of the notebook explores using it to compare against the `pydeseq2` implementation results."],"metadata":{"id":"-Jl57xnAEj4m"}},{"cell_type":"code","source":["# filter genes to those significantly differentially expressed between ground control and space flight\n","df = m.filter_by_dgea(data['255-normalized'], metadata['255'],  pval=0.05, l2fc=0)"],"metadata":{"id":"mLtq8AXyih3P"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["\n","# run DESeq2\n","dds = m.run_deseq2(df, metadata['255'])"],"metadata":{"id":"BxEIiRQeTthJ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["m.sc.tl.pca(dds)\n","m.sc.pl.pca(dds, color='condition', size=200)"],"metadata":{"id":"YB4tKJTDUCym"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**QUESTIONS**\n","\n","1. What is the shape of the dataframe used as input to count the gene expression?  \n","\n","2. How and why is this different than the shape of the original RNA-seq dataframe?\n","\n","3. Do the 2 PCA plots agree in their distribution in 2 dimensions?"],"metadata":{"id":"ULEe4O4yC-Fn"}},{"cell_type":"markdown","source":["**Double click here to enter your answers to the questions above.**\n","\n","1.\n","\n","2.\n","\n","3."],"metadata":{"id":"e_y1hzFPR1sm"}},{"cell_type":"markdown","source":["# Use Gaussian mixture model to cluster the RNA-seq data\n"],"metadata":{"id":"5izcCSuFu-yc"}},{"cell_type":"code","source":["# filter genes to those significantly differentially expressed between ground control and space flight\n","df = m.filter_by_dgea(data['255-normalized'], metadata['255'],  pval=0.05, l2fc=0)"],"metadata":{"id":"Jb8FlBAXvBzq"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# filter data\n","print('shape before filter: ', df.shape)\n","df = m.filter_data(df, dropnans=True, dropgenes=None, droplowcvs=0.1)\n","print('shape after filter: ', df.shape)"],"metadata":{"id":"FbhNG5GWSfvd"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# build gmm with 3 distributions (this may take several minutes with lots of data)\n","gmm = m.my_gmm(df, metadata, 2)"],"metadata":{"id":"VRX8JGFDxCdl"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**QUESTIONS**\n","\n","1. How many genes were dropped from the dataframe after dropping 0.1 or lower CVs?  What do you think the effect will be on the quality of the clustering after dropping so many genes from the data set?\n","\n","2. Does the GMM algorithm require that you specify the number of clusters up-front, or does the algorithm find the ideal number for you?\n","\n","3. Do the GMM cluster results resemble those of k-means?  Why or why not?"],"metadata":{"id":"4nHl22DTR2g8"}},{"cell_type":"markdown","source":["**Double click here to enter your answers to the questions above.**\n","\n","1.\n","\n","2.\n","\n","3."],"metadata":{"id":"kRIQ5L8jR3zC"}},{"cell_type":"markdown","source":["# Plot the RNA-seq gene expression data in a heatmap"],"metadata":{"id":"RefMlqKWmx7M"}},{"cell_type":"code","source":["# filter genes to those significantly differentially expressed between ground control and space flight\n","df = m.filter_by_dgea(data['255-normalized'], metadata['255'],  pval=0.05, l2fc=0)"],"metadata":{"id":"U9WPbWfqtGRF"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# run DESeq2 to get the dds object for plotting the heatmap\n","dds = m.run_deseq2(df, metadata['255'])"],"metadata":{"id":"6DL84PvXoGUq"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# calculate log of count data\n","dds.layers['log1p'] = m.np.log1p(dds.layers['normed_counts'])\n","\n","# map samples to conditions\n"," # transpose df\n","dfT = df.T\n","dfT.columns=dfT.iloc[0]\n","dfT=dfT.iloc[1:]\n","dfT.columns.name=None\n","dfT = dfT.reset_index().rename(columns={\"index\":\"sample\"})\n","conditions = m.map_samples_to_conditions(dfT, metadata['255'], 'Factor Value[Spaceflight]', 'Ground Control', 'Space Flight')\n","\n","# create grapher object for plotting heatmap\n","grapher = m.pd.DataFrame(dds.layers['log1p'].T, index=dds.var_names, columns=list(conditions['sample']))\n","grapher.head()\n","\n","# plot the heatmap\n","m.sns.clustermap(grapher, z_score=0, cmap='RdYlBu_r')"],"metadata":{"id":"fwT3r4mcV0xj"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**QUESTIONS**\n","\n","1.  Which sample is most similar to sample GSM3932698?  How did you choose it?\n","\n","2.  Which gene and sample seem to have the hottest expression?  The coldest?\n","\n","3.  Does the heatmap agree with the PCA plot and k-means clusters?"],"metadata":{"id":"BpBftHe2SfFg"}},{"cell_type":"markdown","source":["**Double click here to enter your answers to the questions above.**\n","\n","1.\n","\n","2.\n","\n","3."],"metadata":{"id":"8l88pNUkR67g"}}]}